cloudinit/config/cc_wireguard.py

```diff
@@ -0,0 +1,243 @@
+            PrivateKey = <private_key>
+            Address = <address>
+            [Peer]
+            PublicKey = <public_key>
+            Endpoint = <endpoint_ip>:<endpoint_ip_port>
+            AllowedIPs = <allowedip1>, <allowedip2>, ...
+        - name: wg1
+          config_path: /etc/wireguard/wg1.conf
+          content: |
+            [Interface]
+            PrivateKey = <private_key>
+            Address = <address>
+            [Peer]
+            PublicKey = <public_key>
+            Endpoint = <endpoint_ip>:<endpoint_ip_port>
+            AllowedIPs = <allowedip1>
+      readinessprobe:
+        - 'systemctl restart service'
+        - 'curl https://webhook.endpoint/example'
+        - 'nc -zv some-service-fqdn 443'
+    """
+        ),
+    ],
+}
+
+__doc__ = get_meta_doc(meta)
+
+LOG = logging.getLogger(__name__)
+
+REQUIRED_WG_INT_KEYS = frozenset(["name", "config_path", "content"])
+
+
+def supplemental_schema_validation(wg_int: dict):
+    """Validate user-provided wg:interfaces option values.
+
+    This function supplements flexible jsonschema validation with specific
+    value checks to aid in triage of invalid user-provided configuration.
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: ValueError describing invalid values provided.
+    """
+    errors = []
+    missing = REQUIRED_WG_INT_KEYS.difference(set(wg_int.keys()))
+    if missing:
+        keys = ", ".join(sorted(missing))
+        errors.append(f"Missing required wg:interfaces keys: {keys}")
+
+    for key, value in sorted(wg_int.items()):
+        if key == "name" or key == "config_path" or key == "content":
+            if not isinstance(value, str):
+                errors.append(
+                    f"Expected a str for wg:interfaces:{key}. Found: {value}"
+                )
+
+    if errors:
+        raise ValueError(
+            f"Invalid wireguard interface configuration:{NL}{NL.join(errors)}"
+        )
+
+
+def write_config(wg_int: dict):
+    """Writing user-provided configuration into Wirguard
+    interface configuration file.
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues writing of configuration file.
+    """
+    LOG.debug("Configuring Wireguard interface %s", wg_int["name"])
+    try:
+        with open(wg_int["config_path"], "w", encoding="utf-8") as wgconfig:
+            wgconfig.write(wg_int["content"])
+    except Exception as e:
+        raise RuntimeError(
+            f"Failure writing Wireguard configuration file:" f"{e}"
+        ) from e
+
+
+def enable_wg(wg_int: dict):
+    """Enable and start Wireguard interface
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues enabling WG interface.
+    """
+    try:
+        LOG.debug("Running: systemctl enable wg-quick@%s", {wg_int["name"]})
+        subp.subp(
+            f'systemctl enable wg-quick@{wg_int["name"]}',
```

Comment:
Drive by comment: Rather than assuming systemd, we should use `cloud.distro.manage_service()`.